Automatically use mediawiki default skin with fallback template

Markus-Rost opened the pull request:
Fix #2213
Fix #2358
Fix #2396

Makes most https://github.com/openzim/mwoffliner/labels/skin issues obsolete

Adds a fallback template based on vector legacy if the specific skin has no template of its own.

Adds a `forceSkin` CLI argument to use a skin supported by the wiki other than the default skin.

kelson42 commented:
We have problems here with the CI. Should be fixed IMHO before review.

Markus-Rost commented:
They are unrelated to the changes here. I've re-run the jobs though.

benoit74 commented on Changelog:
```diff
@@ -6,6 +6,7 @@ Unreleased:
 * CHANGED: Use subtitle from ActionParse (@Markus-Rost #179)
 * NEW: Add progress logs while writing redirects (@benoit74 #2381)
 * FIX: Keep some HTML tags from parser, since they might be used by CSS (especially for backlinks) (@benoit74 #881)
+* NEW: Automatically detect and use mediawiki default skin (@benoit74 #2358)
```
Please add entries for the two other issues as well

benoit74 reviewed (changes requested):
Thank you.

We should:
- add an e2e test case scraping one wiki we know we do not have the skin for (and mention it in the relevant issue so that once we implement this skin we move the e2e test to another wiki ; no need to scrape the whole wiki, only the main page is probably enough
- this test will show that the computation of `this.staticFilesList = getStaticFiles(...` is wrong when using the fallback skin: we do not have a `${skin}.css` resource needed
- add a warning message when we use the fallback skin saying that it is expected the rendering will be a bit "off" ; for instance I did a test on pokemon.fandom.org and it is far from getting a good looking result ; every text is here, readable, but the look is "barebone"

Markus-Rost commented:
> Thank you.
> 
> We should:
> 
> * add an e2e test case scraping one wiki we know we do not have the skin for (and mention it in the relevant issue so that once we implement this skin we move the e2e test to another wiki ; no need to scrape the whole wiki, only the main page is probably enough

I've added a combined test for the `forceSkin` option and the fallback skin by scraping a Wikipedia page with an unsupported skin (`monobook` in this case)

> * this test will show that the computation of `this.staticFilesList = getStaticFiles(...` is wrong when using the fallback skin: we do not have a `${skin}.css` resource needed
> * add a warning message when we use the fallback skin saying that it is expected the rendering will be a bit "off" ; for instance I did a test on pokemon.fandom.org and it is far from getting a good looking result ; every text is here, readable, but the look is "barebone"

fixed that and added the warning as well. Let me know if the warning text is ok or should be more detailed.

Markus-Rost commented on src/renderers/action-parse.renderer.ts:
```diff
@@ -15,18 +16,20 @@ export class ActionParseRenderer extends Renderer {
+      if (['vector', 'vector-2022'].includes(MediaWiki.skin)) {
+        cssResourcesCommon = cssResourcesCommon.concat(MediaWiki.skin)
+      } else {
+        logger.warn(`Unsupported skin ${MediaWiki.skin}, using fallback template to display pages.`)
```
I was considering that, but having the warning inside `templateDeskopArticle()` means it gets logged for every single page inside the ZIM. That seems a bit excessive.

Markus-Rost replied:
But I could probably move the definition of `htmlTemplateCode` into the constructor. It's the same for all pages anyway.

benoit74 reviewed (approved):
LGTM, thank you!

evrial commented:
I think the issue remains
```
docker run --volume=$(pwd)/out:/out -ti ghcr.io/openzim/mwoffliner mwoffliner --verbose --mwUrl=https://www.sigidwiki.com --adminEmail=foo@bar.net --mwActionApiPath=api.php
starting redis-server in the background…
[log] [2025-08-04T19:16:14.280Z] Starting mwoffliner v1.16.1-dev0...
[info] [2025-08-04T19:16:14.284Z] Using custom flavour: no
[log] [2025-08-04T19:16:14.870Z] closing sanitize redis DB
[log] [2025-08-04T19:16:14.871Z] Getting site info...
[log] [2025-08-04T19:16:14.871Z] Getting sub-title...
[info] [2025-08-04T19:16:14.872Z] Getting JSON from [https://www.sigidwiki.com/api.php?action=query&meta=siteinfo&siprop=general%7Cskins%7Crightsinfo&format=json&formatversion=2]
[info] [2025-08-04T19:16:14.874Z] Downloading [https://www.sigidwiki.com/wiki/]
[log] [2025-08-04T19:16:15.042Z] Text direction is [ltr]
[error] [2025-08-04T19:16:15.043Z] FATAL - Failed to get MediaWiki Metadata
[error] [2025-08-04T19:16:15.043Z] Failed to run mwoffliner after [1s]:
 Error: This wiki has no default skin:
[{"code":"metrolook","*":"Metrolook"},{"code":"vector","*":"Vector","unusable":""},{"code":"monobook_old","*":"MonoBook_OLD"},{"code":"monobook","*":"MonoBook","default":""},{"code":"cologneblue","*":"CologneBlue","unusable":""},{"code":"modern","*":"Modern","unusable":""}]
    at MediaWiki.getDefaultSkin (/tmp/mwoffliner/src/MediaWiki.ts:440:13)
    at MediaWiki.getSiteInfo (/tmp/mwoffliner/src/MediaWiki.ts:486:22)
    at process.processTicksAndRejections (node:internal/process/task_queues:105:5)
    at async Promise.all (index 0)
    at async MediaWiki.getMwMetaData (/tmp/mwoffliner/src/MediaWiki.ts:517:116)
    at async Module.execute (/tmp/mwoffliner/src/mwoffliner.lib.ts:194:18)
```

Markus-Rost commented:
The wiki at sigidwiki.com uses MediaWiki 1.22.3, which is below the minimal required version of 1.27

However I do not know how this failed to get caught by the mwMinimalVersion check.

Markus-Rost commented:
> > The wiki at sigidwiki.com uses MediaWiki 1.22.3, which is below the minimal required version of 1.27
> > However I do not know how this failed to get caught by the mwMinimalVersion check.
> 
> I did some black wizardry to see what happens if ... :)

That would explain it :D 

The wiki doesn't have JSON formatversion 2 yet, which would return boolean values as `true` or `false` instead of the XML compatible version of having keys only on true and set to `""`. Therefore mwoffliner can't properly understand a lot of the API responses, like which skin is the default one.

evrial commented:
> The wiki doesn't have JSON formatversion 2 yet, which would return boolean values as `true` or `false` instead of the XML compatible version of having keys only on true and set to `""`. Therefore mwoffliner can't properly understand a lot of the API responses, like which skin is the default one.

Makes sense, so zero chance to make it work?
